File: 48. Rotate Image/main.py
from typing import List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        matrix.reverse()

        for row in range(0, len(matrix)):
            for col in range(row, len(matrix[row])):
                temp = matrix[row][col]
                matrix[row][col] = matrix[col][row]
                matrix[col][row] = temp


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        for row in range(0, len(matrix)):
            for col in range(row, len(matrix[row])):
                temp = matrix[row][col]
                matrix[row][col] = matrix[col][row]
                matrix[col][row] = temp

        for row in range(0, len(matrix)):
            for col in range(0, (int)((len(matrix[row]) + 1) / 2)):
                temp = matrix[row][col]
                matrix[row][col] = matrix[row][len(matrix[row]) - 1 - col]
                matrix[row][len(matrix[row]) - 1 - col] = temp

File: 48. Rotate Image/main.rs
impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        matrix.reverse();

        for row in 0..matrix.len() {
            for col in row..matrix[row].len() {
                let temp = matrix[row][col];
                matrix[row][col] = matrix[col][row];
                matrix[col][row] = temp;
            }
        }
    }
}

impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        for row in 0..matrix.len() {
            for col in row..matrix[row].len() {
                let temp = matrix[row][col];
                matrix[row][col] = matrix[col][row];
                matrix[col][row] = temp;
            }
        }

        for row in 0..matrix.len() {
            let len = matrix[row].len();
            for col in 0..((len + 1) / 2) {
                let temp = matrix[row][col];
                matrix[row][col] = matrix[row][len - 1 - col];
                matrix[row][len - 1 - col] = temp;
            }
        }
    }
}
